### DISCUSSION TITLE
Pretty printing dataclasses

### DISCUSSION THREAD

**Discussion by henryiii** [Textualize/rich#891] [Q&A] [answered]
Very impressed, just started trying it out, so quite new here. When I'm not tethered to an older Python, I love using dataclasses. I printed one out:

<img width="735" alt="Screen Shot 2021-01-08 at 12 26 27 AM" src="https://user-images.githubusercontent.com/4616906/103977921-48742b00-5148-11eb-9a80-bcfcdac2503d.png">

Great colors! But now I'm greedy. I'd like nice spacing too, as that's jumbled and hard to read. Is there an automatic-ish way to get black-like spacing for the output of a data class, without going member by member with something like https://rich.readthedocs.io/en/latest/protocol.html#console-render (adding something is fine, but imagine I have quite a few of these)?

Also, unrelated, adding links to make a new discussion to the issue template chooser is a good way to make discussions a bit more discoverable. :)

**Comment by willmcgugan** [comment, marked as answer]
Yes, it would be nice to format Dataclasses similar to dicts. I did try, but it is more complex to introspect a dataclass than a dict. I might try again at some point.

> Also, unrelated, adding links to make a new discussion to the issue template chooser is a good way to make discussions a bit more discoverable. :)

Good idea.

**Reply by henryiii** [reply]
This is embarrassing, I'm just now looking over my code and realized I'm using attrs instead. There must have been something I needed that dataclasses didn't do. Anyway, this is what I came up with. To make it do this by default, you'd want to check to see if a dataclass had a custom repr/str, which might be hard, and you'd use dataclasses.asdict instead of attr.asdict. (and you wouldn't want it to be a table, but this was easy and looks fine for my use).

```python
    def __rich_console__(self, console: Console, options: ConsoleOptions) -> RenderResult:
        yield f"[b]{self.__class__.__name__}:[/b]"
        my_table = Table("Attribute", "Value")
        for k, v in attr.asdict(self).items():
            my_table.add_row(k, Syntax(repr(v), "python"))
        yield my_table

```

If I end up using the output, I might try to touch it up further, like trying out the URL feature. :)

PS, unrelated: `-m rich.markdown` is amazing; perfect addition to `bat`. :)

**Reply by willmcgugan** [reply]
That's a reasonable solution. You could also render_scope as an alternative. SOmething like this (untested).

```python
from rich.scope import render_scope

def __rich_console__(self, console: Console, options: ConsoleOptions) -> RenderResult:
    yield render_scope(attr.asdict(self), title=self.__class__.__name__)
```

Ideally though I'd like something which integrates with pretty.py so it can handle nesting better. i.e. a dataclass containing a reference to another dataclass.
